File: 20.QuickSort.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )                                         */
/*  ___ _                      _              _                                 */
/*  | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)                                */
/*  | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |                                */
/*  |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |                                */
/*                                     |__/                                     */
/* ----------------------------------------------------------------------------  */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                         */
/*    Github : https://github.com/Manu577228                                    */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio         */
/* ----------------------------------------------------------------------------  */

import java.io.*;

public class QuickSortExample {

    // Partition function (Lomuto partition scheme)
    static int partition(int[] a, int low, int high) {
        int pivot = a[high];                  // choose last element as pivot
        int i = low - 1;                      // i will track position of smaller elements
        for (int j = low; j < high; j++) {    // iterate from low to high-1
            if (a[j] <= pivot) {              // if current element <= pivot, place in left side
                i++;
                int temp = a[i];              // swap a[i] and a[j]
                a[i] = a[j];
                a[j] = temp;
            }
        }
        int temp = a[i + 1];                  // place pivot after last smaller element
        a[i + 1] = a[high];
        a[high] = temp;
        return i + 1;                         // return pivot index
    }

    // QuickSort recursive function
    static void quicksort(int[] a, int low, int high) {
        if (low < high) {                     // base case: at least two elements
            int p = partition(a, low, high);  // partition the array and get pivot index
            quicksort(a, low, p - 1);         // recursively sort left partition
            quicksort(a, p + 1, high);        // recursively sort right partition
        }
    }

    public static void main(String[] args) throws IOException {
        int[] arr = {3, 6, 8, 10, 1, 2, 1};   // initial unsorted list

        System.out.print("Original array: ");
        for (int v : arr) System.out.print(v + " ");
        System.out.println();

        quicksort(arr, 0, arr.length - 1);    // call quicksort

        System.out.print("Sorted array: ");
        for (int v : arr) System.out.print(v + " ");
        System.out.println();
    }
}

File: 20.QuickSort.py
# Quick Sort is an in-place, divide-and-conquer sorting algorithm that picks a 
# pivot and partitions the array so values less than the pivot come before it and values greater come after. 
# It then recursively sorts the left and right partitions until the whole array is sorted.

# Explanation

# Divide and conquer: Quick Sort divides the problem (array) into two subproblems (left and right of the pivot) 
# by partitioning around a pivot element.

# Partition step: Choose a pivot (common simple choice: last element). 
# Rearrange the array so elements ≤ pivot go left, and elements > pivot go right — then place the pivot between them.

# Recursion: Recursively apply quick sort to left and 
# right subarrays (excluding the pivot which is now in correct place).

# Complexity: Average time O(n log n), worst-case O(n²) (when pivot choices 
# are poor, e.g., already sorted input and picking last element), space O(log n) average for recursion stack (O(n) worst-case).

# In-place but not stable: The common in-place Lomuto/Hoare 
# partitions sort without extra arrays but do not preserve relative order of equal elements (not stable).

# Pivot choices & optimizations: Use randomized pivot or 
# median-of-three to avoid worst-case patterns; switch to insertion sort for small subarrays; 
# tail recursion elimination can reduce stack use.

arr = [3, 6, 8, 10, 1, 2, 1];

print("original Array:", arr)

def partition(a, low, high):
    pivot = a[high]
    i = low - 1
    for j in range(low, high):
        if a[j] <= pivot:
            i += 1
            a[i], a[j] = a[j], a[i]
    a[i + 1], a[high] = a[high], a[i + 1]
    return i + 1
    
def quicksort(a, low, high):
    if low < high:
        p = partition(a, low, high)
        quicksort(a, low, p - 1)
        quicksort(a, p + 1, high)

quicksort(arr, 0, len(arr) - 1)

print("Sorted Array: ", arr)

